packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -170,6 +200,75 @@ library LibProposing {
         if (blk.blockId != blockId) revert L1_BLOCK_ID();
     }
 
+    function _assignProver(
+        TaikoData.State storage state,
+        TaikoData.Config memory config,
+        bytes32 metaHash,
+        uint16 proofWindow,
+        uint32 gasLimit,
+        uint64 blockId,
+        address prover,
+        uint32 maxFeePerGas,
+        bytes memory proverParams
+    )
+        private
+        returns (address _actualProver, uint32 _feePerGas, uint64 _bond)
+    {
+        if (prover == address(0)) {
+            _actualProver = prover;
+            _feePerGas = maxFeePerGas;
+        } else if (!prover.isContract()) {
+            // Verify the prover has authorized this assignment
+            bytes32 hash =
+                keccak256(abi.encodePacked("PROVE_TAIKO_BLOCK", metaHash));
```

Comment:
Since `metaHash` contains: `block.timestamp` \ `block.number` \ `block.prevrandao` \ `state.slotB.numBlocks`, seems its impossible for a user to sign this payload beforehand?

Reply:
True...

Reply:
Normally the proposer will know it because they will only want to propose blocks that they know exactly how they will behave so they can optimize for the transactions they include in the block. And so because the proposer knows, they will only request a quote from provers after they are done building, it makes sense to also give it to provers so they will also know exactly what block they will have to prove so they can accurately predict their proving costs (or at least more accurately, they don't know how many more blocks they will get to prove and what the Ethereum gas cost will be).

But I totally agree for now it is safer not to include it because proposers will unlikely be very advanced initially and may be very problematic on testnets. In the future I think having the option for the prover to sign off on the exact block data as an option may be useful.